### PR TITLE
simplify RC and SVC listers

### PR DESCRIPTION
**Pull request by deads2k**
Make the RC and SVC listers use the common list functions that more closely match client APIs, are consistent with other listers, and avoid unnecessary copies.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32888)

### CODE REVIEW

File: pkg/client/cache/listers.go [hunk 1]
```diff
@@ -31,104 +32,66 @@ import (
...
-// Example:
-// s := cache.NewStore()
-// lw := cache.ListWatch{Client: c, FieldSelector: sel, Resource: "pods"}
-// r := cache.NewReflector(lw, &api.Pod{}, s).Run()
-// l := StoreToPodLister{s}
-// l.List()
-type StoreToPodLister struct {
-	Indexer
-}
+type AppendFunc func(interface{})
```
Comment (smarterclayton) [inline pkg/client/cache/listers.go]: Godoc

File: pkg/client/cache/listers_core.go [hunk 2]
```diff
@@ -71,3 +73,72 @@ func (s storePodsNamespacer) Get(name string) (*api.Pod, error) {
 	}
 	return obj.(*api.Pod), nil
 }
+
+/*****************************************************************************/
```
Comment (smarterclayton) [inline pkg/client/cache/listers_core.go]: Unusual, we don't do this.  Remove.  Add nice godoc to StoreToReplicationControllerLister instead.

File: pkg/client/cache/listers_core.go [hunk 3]
```diff
@@ -142,3 +142,66 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
 	}
 	return
 }
+
+/*****************************************************************************/
```
Comment (smarterclayton) [inline pkg/client/cache/listers_core.go]: Unusual remove

**Review by smarterclayton — CHANGES_REQUESTED** [review]
Minor changes, godoc on the lister types.

**Comment by deads2k** [comment]
@smarterclayton what about the changes from shallow copies to pointers?

**Comment by deads2k** [comment]
comments addressed.

**Comment by deads2k** [comment]
@k8s-bot test this: issue #31706

**Comment by smarterclayton** [comment]
Pointers are better than shallow copies in almost every scenario except an specific optimization or for gross user simplicity.

**Comment by deads2k** [comment]
@smarterclayton any other changes or are you good with this?

**Comment by deads2k** [comment]
in person lgtm from clayton

**Review by smarterclayton — APPROVED** [review]

File: pkg/client/cache/listers_core.go [hunk 4]
```diff
@@ -71,3 +73,133 @@ func (s storePodsNamespacer) Get(name string) (*api.Pod, error) {
...
+}
+
+func (s storeReplicationControllersNamespacer) List(selector labels.Selector) (ret []*api.ReplicationController, err error) {
+	err = ListAllByNamespace(s.indexer, s.namespace, selector, func(m interface{}) {
+		ret = append(ret, m.(*api.ReplicationController))
+	})
+	return ret, err
+}
+
+func (s storeReplicationControllersNamespacer) Get(name string) (*api.ReplicationController, error) {
```
Comment (sttts) [inline pkg/client/cache/listers_core.go]: any reason to have a non-pointer receiver here and above?

Comment (deads2k) [inline pkg/client/cache/listers_core.go]: > any reason to have a non-pointer receiver here and above?

No, cargo-cult.

File: pkg/client/cache/listers_core.go [hunk 5]
```diff
@@ -141,3 +141,65 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
...
 }
+
+// StoreToServiceLister helps list services
+type StoreToServiceLister struct {
+	Indexer Indexer
+}
+
+func (s *StoreToServiceLister) List(selector labels.Selector) (ret []*api.Service, err error) {
+	err = ListAll(s.Indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(*api.Service))
```
Comment (sttts) [inline pkg/client/cache/listers_core.go]: this potentially returns non-nil lists for `err != nil`. Intentional?

File: pkg/client/cache/listers_core.go [hunk 6]
```diff
@@ -141,3 +141,65 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
...
+
+type storeServicesNamespacer struct {
+	indexer   Indexer
+	namespace string
+}
+
+func (s storeServicesNamespacer) List(selector labels.Selector) (ret []*api.Service, err error) {
+	err = ListAllByNamespace(s.indexer, s.namespace, selector, func(m interface{}) {
+		ret = append(ret, m.(*api.Service))
+	})
```
Comment (sttts) [inline pkg/client/cache/listers_core.go]: here again, ret != nil for err != nil.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 7]
```diff
@@ -670,7 +670,7 @@ func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, n
 	// skip looking at other pods in the service if the current pod defines all the required affinity labels
 	if !labelsExist {
 		services, err := s.serviceLister.GetPodServices(pod)
-		if err == nil {
+		if err == nil && len(services) > 0 {
```
Comment (sttts) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: why this additional check? Err should be non-nil in that case.

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 8]
```diff
@@ -194,7 +194,7 @@ func NewServiceAntiAffinityPriority(podLister algorithm.PodLister, serviceLister
 // The label to be considered is provided to the struct (ServiceAntiAffinity).
 func (s *ServiceAntiAffinity) CalculateAntiAffinityPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodes []*api.Node) (schedulerapi.HostPriorityList, error) {
 	var nsServicePods []*api.Pod
-	if services, err := s.serviceLister.GetPodServices(pod); err == nil {
+	if services, err := s.serviceLister.GetPodServices(pod); err == nil && len(services) > 0 {
```
Comment (sttts) [inline plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: Like above, err == nil => len(services) > 0

**Pull request merged**